sklearn/inspection/permutation_importance.py

```diff
@@ -0,0 +1,97 @@
+from ..utils.validation import check_array
+from ..metrics import check_scoring
+
+
+def _calculate_permutation_scores(estimator, X, y, col_idx, random_state,
+                                  n_rounds, scorer):
+    """Calcuate permutation scores for a column."""
+    X = X.copy()
+    if hasattr(X, 'iloc'):  # pandas dataframe
+        X_iloc = X.iloc
+    else:
+        X_iloc = X
+
+    scores = np.zeros(n_rounds)
+    temp = X_iloc[:, col_idx].values.copy()
+    for n_round in range(n_rounds):
+        random_state.shuffle(temp)
+        X_iloc[:, col_idx] = temp
+        feature_score = scorer(estimator, X, y)
+        scores[n_round] = feature_score
+    return scores
+
+
+def permutation_importance(estimator, X, y, scoring=None, n_rounds=1,
+                           n_jobs=None, random_state=None):
+    """Permutation importance for feature evaluation. [BRE]_
+
+    The permutation importance of a feature is calculated as follows. First,
+    the estimator is trained on a training set. Then a baseline metric, defined
+    by ``scoring``, is evaluated on a (potentially different) dataset defined
+    by the ``X`` parameter. Next, a feature column from the validation set is
+    permuted and the metric is evaluated again. The permutation importance is
+    defined to be the difference between the baseline metric and metric from
+    permutating the feature column.
+
+    Read more in the :ref:`User Guide <permutation_importance>`.
+
+    Parameters
+    ----------
+    estimator : object
+        An estimator that has already been `fit` and is compatible with
+        ``scorer``.
+
+    X : array-like or DataFrame, shape = (n_samples, n_features)
+        Data on which permutation importance will be computed.
+
+    y : array-like, shape = (n_samples, ...)
+        Targets for supervised learning.
+
+    scoring : string, callable or None, optional (default=None)
+        Scorer to use. It can be a single
+        string (see :ref:`scoring_parameter`) or a callable (see
+        :ref:`scoring`). If None, the estimator's default scorer is used.
+
+    n_rounds : int, optional (default=1)
+        Number of times to permute a feature.
+
+    n_jobs : int or None, optional (default=None)
+        The number of jobs to use for the computation.
+        ``None`` means 1 unless in a :obj:`joblib.parallel_backend` context.
+        ``-1`` means using all processors. See :term:`Glossary <n_jobs>`
+        for more details.
+
+    random_state : int, RandomState instance or None, optional, default None
+        Pseudo-random number generator to control the permutations.
+        See :term:`random_state`.
+
+    Returns
+    -------
+    importances : array, shape (n_features, n_rounds)
+        Permutation importance scores.
+
+    References
+    ----------
+    .. [BRE] L. Breiman, "Random Forests", Machine Learning, 45(1), 5-32,
+        2001.https://doi.org/10.1023/A:1010933404324
+
+    """
+    if not hasattr(X, 'iloc'):
```

Comment:
why don't we just call ``check_array`` and convert? To keep feature names?
I would really try to keep pandas stuff out of this for now.

Reply:
`check_array` will always return a numpy array.

I am thinking of the use case where the estimator is a pipeline with a column transformer.

Reply:
yes that was my point ;)
Hm ok that's fair. You could probably use the ``safe_indexing`` helper, right?

Reply:
In `_calculate_permutation_scores`, we are accessing and then mutating the dataframe. This would require something like a `safe_setting_indexing(df, index, numpy_array)`.